### PR TITLE
Databricks - explicit networking dependencies and NSG rules

### LINKED ISSUE #141: CI build Databricks (and other) transient errors
We are seeing several transient errors which we will record here for tracking. We should consider finding root cause resolution and failing that use [Terragrunt error hooks](https://terragrunt.gruntwork.io/docs/features/hooks/) to retry them automatically.

### Cannot create cluster
```
Error: cannot create cluster: 0303-131037-qid874qx is not able to transition from TERMINATED to RUNNING: Self-bootstrap failure during launch. Please try again later and contact Databricks if the problem persists. databricks_error_message: Self-bootstrap failure during launch. Please try again later a..., Termination info: code: SELF_BOOTSTRAP_FAILURE, type: , parameters: map[databricks_error_message:Self-bootstrap failure during launch. Please try again later and contact Databricks if the problem persists. Node daemon fast failed and did not answer ping for instance 04387a5e9e4a4328b335937bb7db2e10]. Please see https://docs.databricks.com/dev-tools/api/latest/clusters.html#clusterclusterstate for more details
[2081](https://github.com/jjgriff93/FlowEHR/actions/runs/4323583714/jobs/7547534079#step:3:2082)
│ 
[2082](https://github.com/jjgriff93/FlowEHR/actions/runs/4323583714/jobs/7547534079#step:3:2083)
│   with databricks_cluster.fixed_single_node,
[2083](https://github.com/jjgriff93/FlowEHR/actions/runs/4323583714/jobs/7547534079#step:3:2084)
│   on main.tf line 53, in resource "databricks_cluster" "fixed_single_node":
[2084](https://github.com/jjgriff93/FlowEHR/actions/runs/4323583714/jobs/7547534079#step:3:2085)
│   53: resource "databricks_cluster" "fixed_single_node" {
```

Getting this often in new deployments to `westeurope`

***
+1 on the above error (uksouth)

### PR DESCRIPTION
**Pull request by jjgriff93**
# Resolves #141 

## What is being addressed

The saga continues. Still seeing the occasional failure - this is the very last theory to test before giving up and implementing some retry logic in Terragrunt. The support engineer can't find any issues with our set-up.

This PR adds some more explicit dependencies to the cluster creation to ensure every networking component is in place before proceeding. We also manually add the NSG rules that are added automatically by Databricks service delegation as there is potentially a delay in these rules being added - other examples of databricks terraform implementations seem to do this, for reasons unknown but that's my assumption. Also have bumped the databricks provider to the latest version.

Fingers crossed again.

### CODE REVIEW

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4578564946 (with refid `afef6614`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/232#issuecomment-1492497594) from @jjgriff93)

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4578689471 (with refid `afef6614`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/232#issuecomment-1492530641) from @jjgriff93)

File: infrastructure/transform/network.tf [hunk 1]
```diff
@@ -17,6 +17,85 @@ resource "azurerm_network_security_group" "databricks" {
   resource_group_name = var.core_rg_name
   location            = var.core_rg_location
   tags                = var.tags
+
+  # Add NSG rules manually to ensure they are in place before deploying a cluster instead of waiting for service delegation
```
Comment (t-young31) [inline infrastructure/transform/network.tf:21]: is there a way to de-duplicate this? (assuming the dbks NSG still gets deployed)

**Review by t-young31 — APPROVED** [review]
once again: 🤞🏼

**Review by damoodamoo — APPROVED** [review]

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4595454497 (with refid `afef6614`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/232#issuecomment-1494046436) from @jjgriff93)

**Comment by jjgriff93** [comment]
Tests just hit the bootstrap error again, so this hasn't solved it. Will abandon as it introduces unnecessary complexity

**Pull request closed without merging**